File: ui/core/wasm/bulk_sim/progress.ts
import { BulkSimStage, BulkSimStageMetrics, ProgressMetrics } from '../../proto/api';
import { WorkerProgressCallback } from '../../worker_pool';
import { ConcurrentBulkSimStageConfig } from './types';

export const bulkSimStageLogName = (stage: BulkSimStage): string => {
	switch (stage) {
		case BulkSimStage.BulkSimStageLow:
			return 'low';
		case BulkSimStage.BulkSimStageMedium:
			return 'medium';
		case BulkSimStage.BulkSimStageHigh:
			return 'high';
		default:
			return BulkSimStage[stage] ?? String(stage);
	}
};

export const formatBulkSimStageStart = (
	config: ConcurrentBulkSimStageConfig,
	candidateCount: number,
	minIterations: number,
	carriedIterations: number,
): string => {
	return `- Stage: ${bulkSimStageLogName(config.stage)} - Starting
Sims:
  Candidates: ${candidateCount}
  Total runs: ${candidateCount + 1}-${candidateCount + 2} (baseline probe, optional baseline, candidates)
Stage config:
  Per-candidate concurrent sim: true
  Min iterations: ${minIterations}
  Carried iterations: ${carriedIterations}
  Target error: ${config.targetErrorPct.toFixed(2)}%`;
};

export const formatBulkSimStageSummary = (status: string, metrics: BulkSimStageMetrics, completedSims: number): string => {
	return `- Stage: ${bulkSimStageLogName(metrics.stage)} - ${status}
Sims:
  Input gear sets: ${metrics.inputGearSets}
  Completed candidates: ${completedSims}
  Survivors: ${metrics.survivors}
Results:
  Iterations: ${metrics.iterations}
  Target error: ${metrics.targetErrorPct.toFixed(2)}%
  Observed error: ${metrics.observedErrorPct.toFixed(2)}%
  Best candidate DPS: ${metrics.bestCandidateAvgDps.toFixed(2)}
  Baseline DPS: ${metrics.baselineAvgDps.toFixed(2)}
Timing:
  Duration: ${metrics.durationSeconds.toFixed(2)}s`;
};

// Binds the parts of a stage's progress reporting that never change between emits, so
// call sites pass only what actually moves. Stage totals are only known after the
// baseline probe, so a stage uses one emitter for the probe and another for the rest.
export type BulkSimStageProgressEmitter = {
	report: (completedSims: number, completedIterations: number, dps: number) => void;
};

export const makeBulkSimStageProgressEmitter = (
	onProgress: WorkerProgressCallback,
	stage: BulkSimStage,
	totalSims: number,
	totalIterations: number,
): BulkSimStageProgressEmitter => ({
	report: (completedSims, completedIterations, dps) =>
		emitBulkSimStageProgress(onProgress, stage, completedSims, totalSims, completedIterations, totalIterations, dps),
});

const emitBulkSimStageProgress = (
	onProgress: WorkerProgressCallback,
	bulkStage: BulkSimStage,
	completedSims: number,
	totalSims: number,
	completedIterations: number,
	totalIterations: number,
	dps: number,
) => {
	onProgress(
		ProgressMetrics.create({
			bulkStage,
			completedSims,
			totalSims,
			completedIterations,
			totalIterations,
			dps,
		}),
	);
};

File: sim/core/bulk/progress.go
package bulk

import (
	"fmt"
	"sync"
	"time"

	"github.com/wowsims/tbc/sim/core/proto"
)

type BulkSimStageProgressTracker struct {
	mutex                               sync.Mutex
	emitter                             bulkSimStageProgressEmitter
	totalCandidates                     int
	iterations                          int32
	completedSimsBeforeCandidates       int
	completedIterationsBeforeCandidates int32
	completedCandidates                 int
	completedCandidateIterations        int32
	completedIterationsByCandidate      []int32
	lastProgressEmit                    time.Time
}

// Binds the parts of a stage's progress reporting that never change between emits, so
// call sites pass only what actually moves. Stage totals are only known after the
// baseline probe, so a stage uses one emitter for the probe and another for the rest.
type bulkSimStageProgressEmitter struct {
	progress        chan *proto.ProgressMetrics
	stage           proto.BulkSimStage
	totalSims       int
	totalIterations int32
}

func (emitter bulkSimStageProgressEmitter) report(completedSims int, completedIterations int32, dps float64) {
	emitBulkSimStageProgress(emitter.progress, emitter.stage, completedSims, emitter.totalSims, completedIterations, emitter.totalIterations, dps)
}

func emitBulkSimStageProgress(progress chan *proto.ProgressMetrics, stage proto.BulkSimStage, completedSims int, totalSims int, completedIterations int32, totalIterations int32, dps float64) {
	if progress == nil {
		return
	}

	progress <- &proto.ProgressMetrics{
		BulkStage:           stage,
		CompletedSims:       int32(completedSims),
		TotalSims:           int32(totalSims),
		CompletedIterations: completedIterations,
		TotalIterations:     totalIterations,
		Dps:                 dps,
	}
}

func (tracker *BulkSimStageProgressTracker) reportCandidateProgress(position int, progressMetrics *proto.ProgressMetrics) {
	if progressMetrics == nil || progressMetrics.TotalIterations == 0 {
		return
	}
	tracker.report(position, min(progressMetrics.CompletedIterations, tracker.iterations), progressMetrics.Dps, false)
}

func (tracker *BulkSimStageProgressTracker) reportCandidateComplete(position int, result *BulkSimCandidateResult) {
	dps := 0.0
	if result != nil && result.DpsMetrics != nil {
		dps = result.DpsMetrics.Avg
	}
	tracker.report(position, tracker.iterations, dps, true)
}

// Records how far one candidate has got and emits stage progress when an update is due.
// A finished candidate always emits, since sim counts are what the caller is waiting on.
func (tracker *BulkSimStageProgressTracker) report(position int, completedIterations int32, dps float64, candidateFinished bool) {
	if tracker.emitter.progress == nil || position < 0 || position >= tracker.totalCandidates {
		return
	}

	tracker.mutex.Lock()
	shouldEmit := candidateFinished || tracker.shouldEmitProgressLocked()
	if completedIterations > tracker.completedIterationsByCandidate[position] {
		tracker.completedCandidateIterations += completedIterations - tracker.completedIterationsByCandidate[position]
		tracker.completedIterationsByCandidate[position] = completedIterations
		shouldEmit = shouldEmit || tracker.shouldEmitProgressLocked()
	}
	if candidateFinished {
		tracker.completedCandidates++
	}
	totalCompletedIterations := tracker.completedIterationsBeforeCandidates + tracker.completedCandidateIterations
	completedSims := tracker.completedSimsBeforeCandidates + tracker.completedCandidates
	if shouldEmit {
		tracker.lastProgressEmit = time.Now()
	}
	tracker.mutex.Unlock()
	if !shouldEmit {
		return
	}

	tracker.emitter.report(completedSims, totalCompletedIterations, dps)
}

func (tracker *BulkSimStageProgressTracker) shouldEmitProgressLocked() bool {
	return tracker.lastProgressEmit.IsZero() || time.Since(tracker.lastProgressEmit) >= BulkSimProgressThrottle
}

func setBulkSimStageTiming(timings *proto.BulkSimTimings, stage proto.BulkSimStage, durationSeconds float64) {
	switch stage {
	case proto.BulkSimStage_BulkSimStageLow:
		timings.LowStageSeconds = durationSeconds
	case proto.BulkSimStage_BulkSimStageMedium:
		timings.MediumStageSeconds = durationSeconds
	case proto.BulkSimStage_BulkSimStageHigh:
		timings.HighStageSeconds = durationSeconds
	}
}

func formatBulkSimStageStart(config BulkSimStageConfig, candidateCount int, concurrency int, minIterations int32, carriedIterations int32) string {
	return fmt.Sprintf("- Stage: %s - Starting\n"+
		"Sims:\n"+
		"  Candidates: %d\n"+
		"  Total runs: %d-%d (baseline probe, optional baseline, candidates)\n"+
		"Stage config:\n"+
		"  Concurrency: %d\n"+
		"  Per-candidate concurrent sim: %t\n"+
		"  Min iterations: %d\n"+
		"  Carried iterations: %d\n"+
		"  Target error: %.2f%%",
		bulkSimStageLogName(config.Stage),
		candidateCount,
		candidateCount+1,
		candidateCount+2,
		concurrency,
		config.UseConcurrentSim,
		minIterations,
		carriedIterations,
		config.TargetErrorPct,
	)
}

func formatBulkSimStageSummary(status string, metrics *proto.BulkSimStageMetrics, completedSims int) string {
	return fmt.Sprintf("- Stage: %s - %s\n"+
		"Sims:\n"+
		"  Input gear sets: %d\n"+
		"  Completed candidates: %d\n"+
		"  Survivors: %d\n"+
		"Results:\n"+
		"  Iterations: %d\n"+
		"  Target error: %.2f%%\n"+
		"  Observed error: %.2f%%\n"+
		"  Best candidate DPS: %.2f\n"+
		"  Baseline DPS: %.2f\n"+
		"Timing:\n"+
		"  Duration: %.2fs",
		bulkSimStageLogName(metrics.Stage),
		status,
		metrics.InputGearSets,
		completedSims,
		metrics.Survivors,
		metrics.Iterations,
		metrics.TargetErrorPct,
		metrics.ObservedErrorPct,
		metrics.BestCandidateAvgDps,
		metrics.BaselineAvgDps,
		metrics.DurationSeconds,
	)
}

func bulkSimStageLogName(stage proto.BulkSimStage) string {
	switch stage {
	case proto.BulkSimStage_BulkSimStageLow:
		return "low"
	case proto.BulkSimStage_BulkSimStageMedium:
		return "medium"
	case proto.BulkSimStage_BulkSimStageHigh:
		return "high"
	default:
		return stage.String()
	}
}
